gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java

```diff
@@ -1005,6 +1005,162 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
+      final LogWriter logger, boolean throwErrorOnMismatch) {
+    try {
+      for (GemFireContainer container : uninitializedContainers) {
+        LocalRegion region = container.getRegion();
+        DataPolicy dp = region.getDataPolicy();
+        if (dp == DataPolicy.PERSISTENT_PARTITION || dp == DataPolicy.PERSISTENT_REPLICATE) {
+          GfxdIndexManager gim = (GfxdIndexManager)region.getIndexUpdater();
+          if (gim == null || container.isGlobalIndex()) continue;
+          int localRegionSz = getLocalRegionSz(region, dp, logger, false, throwErrorOnMismatch);
+          List<GemFireContainer> allIndexes = gim.getAllIndexes();
+          for (GemFireContainer c : allIndexes) {
+            if (c.isLocalIndex()) {
+              if (c.getIndexSize() != localRegionSz) {
+                if (!throwErrorOnMismatch) {
+                  logger.warning("checkRecoveredIndex: for table: " + region.getName() + " " +
+                      "number of local entries = " + localRegionSz + " and number of " +
+                      "index entries in the index: " + c.getName() + " = " + c.getIndexSize());
+                }
+                else {
+                  logger.error("checkRecoveredIndex: for table: " + region.getName() + " " +
+                      "number of local entries = " + localRegionSz + " and number of " +
+                      "index entries in the index: " + c.getName() + " = " + c.getIndexSize());
+                  dumpIndexAndRegion(region, dp, c, logger);
+                  throw new IllegalStateException("Table data and indexes are not reconciling." +
+                      "Probably need to revoke the disk store");
+                }
+                logger.info("FabricDatabase: index and region out of sync. Recreating the indexes");
+                // First clear all the indexes
+                clearAllIndexes(uninitializedContainers);
+                recreateAllLocalIndexes(logger);
+                checkRecoveredIndex(uninitializedContainers, logger, true);
+              } else {
+                if (logger.fineEnabled()) {
+                  logger.fine("checkRecoveredIndex: local index: " + c.getName() +
+                      " and table: " + region.getName() + " with size: " + localRegionSz);
+                }
+              }
+            }
+          }
+        }
+      }
+    } finally {
+      // clear the init recovery map
+      Collection<DiskStoreImpl> diskStores = Misc.getGemFireCache().listDiskStores();
+      for (DiskStoreImpl ds : diskStores) {
+        if (!ds.getName().equals(GfxdConstants.GFXD_DD_DISKSTORE_NAME)) {
+          ds.clearRecoveryMapOfAllOplogs();
+        }
+      }
+    }
+  }
+
+  private void dumpIndexAndRegion(LocalRegion region, DataPolicy dp, GemFireContainer index, LogWriter logger) {
+    ((MemIndex)index.getConglomerate()).dumpIndex("Dumping all indexes");
+    getLocalRegionSz(region, dp, logger, true, false);
```

Comment:
Since this method is also dumping some info, better to name it something like getAndDumpLocalRegionSize?